### PR TITLE
Add DELTA_BYTE_ARRAY encoder for Parquet

### PR DESCRIPTION
**Pull request by etseidl**
## Description
Re-submission of #14938. Final (delta) piece of #13501.

Adds the ability to encode Parquet pages as DELTA_BYTE_ARRAY. Python testing wlll be added as a follow-on when per-column encoding selection is added to the python API (ref this [comment](https://github.com/rapidsai/cudf/pull/15081#issuecomment-1979731930)).

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by etseidl** [comment]
cc @vuule @PointKernel @nvdbaranec

**Review by vuule — COMMENTED** [review]
Skipped the parts that haven't changed from https://github.com/rapidsai/cudf/pull/14938

File: cpp/tests/io/parquet_reader_test.cpp [hunk 1]
```diff
@@ -1999,9 +2000,29 @@ TEST_F(ParquetReaderTest, DeltaSkipRowsWithNulls)
...
+  input_metadata.column_metadata[18].set_encoding(column_encoding::DELTA_LENGTH_BYTE_ARRAY);
+  input_metadata.column_metadata[19].set_encoding(column_encoding::DELTA_LENGTH_BYTE_ARRAY);
+  input_metadata.column_metadata[20].set_encoding(column_encoding::DELTA_BYTE_ARRAY);
+  input_metadata.column_metadata[21].set_encoding(column_encoding::DELTA_BYTE_ARRAY);
+  input_metadata.column_metadata[22].set_encoding(column_encoding::DELTA_BYTE_ARRAY);
+  input_metadata.column_metadata[23].set_encoding(column_encoding::DELTA_BYTE_ARRAY);
+  input_metadata.column_metadata[24].set_encoding(column_encoding::DELTA_BYTE_ARRAY);
+  input_metadata.column_metadata[25].set_encoding(column_encoding::DELTA_BYTE_ARRAY);
+  input_metadata.column_metadata[26].set_encoding(column_encoding::DELTA_BYTE_ARRAY);
+  input_metadata.column_metadata[27].set_encoding(column_encoding::DELTA_BYTE_ARRAY);
```
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/tests/io/parquet_reader_test.cpp:2024]: Yeah, was thinking that too. Will do.

[the remark this answers is not in the document]
Comment (vuule) [reply cpp/tests/io/parquet_reader_test.cpp:2024]:
```
for (int i = 12; i <= 27; ++i) {
    input_metadata.column_metadata[i].set_encoding(
        i <= 19 ? column_encoding::DELTA_LENGTH_BYTE_ARRAY : column_encoding::DELTA_BYTE_ARRAY
    );
}
```

File: cpp/src/io/parquet/page_enc.cu [hunk 2]
```diff
@@ -493,10 +495,47 @@ CUDF_KERNEL void __launch_bounds__(128)
...
+{
+  // first check for dictionary (boolean always uses dict encoder)
+  if (chunk->use_dictionary or col_desc->physical_type == BOOLEAN) {
+    return encode_kernel_mask::DICTIONARY;
+  }
+
+  // next check for user requested encoding, but skip if user requested dictionary encoding
+  // (if we could use the requested dict encoding, we'd have returned above)
+  if (col_desc->requested_encoding != column_encoding::USE_DEFAULT and
+      col_desc->requested_encoding != column_encoding::DICTIONARY) {
```
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/page_enc.cu:511]: That will have been checked on the host side. If dict cannot be used even if requested, then use_dictionary will be set to false.

File: cpp/tests/io/parquet_reader_test.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/tests/io/parquet_reader_test.cpp:2024]: You saw the future where I used two loops 😂

[the remark this answers is not in the document]
Comment (vuule) [reply cpp/tests/io/parquet_reader_test.cpp]: That would be fine as well.
I just experimented with Copilot and it used a single loop. So I thought I might as well share the code that it generated.

**Review by vuule — APPROVED** [review]

File: cpp/src/io/parquet/page_enc.cu [hunk 3]
```diff
@@ -3137,6 +3429,13 @@ void EncodePages(device_span<EncPage> pages,
     gpuEncodeDeltaLengthByteArrayPages<encode_block_size>
       <<<num_pages, encode_block_size, 0, strm.value()>>>(pages, comp_in, comp_out, comp_results);
   }
+  if (BitAnd(kernel_mask, encode_kernel_mask::DELTA_BYTE_ARRAY) != 0) {
```
Comment (PointKernel) [inline cpp/src/io/parquet/page_enc.cu:3432]: I've seen this utility in multiple places and just realized it's just `cuda::std::bit_and()` defined in `<cuda/std/functional>`.

could be a separate PR removing this also `BitOr`

**Review by PointKernel — APPROVED** [review]
LGTM :fire:

File: cpp/src/io/parquet/page_enc.cu (hunk 3 above)
Comment (etseidl) [reply cpp/src/io/parquet/page_enc.cu:3432]: There's a slight difference in that `BitAnd` allows either argument to be a scoped enum. Since the scoped enums can't be implicitly cast to an int, I don't think the `std::bit_and()` will be happy. I was trying to avoid `static_cast<uint32_t>` all over the place, and @vuule _really_ didn't want me overloading 'operator &()` :joy:. But we can revisit this.

Comment (PointKernel) [reply cpp/src/io/parquet/page_enc.cu:3432]: Oh my... ignore me then. I'm happy with the current the way in this case.
